File: Codility/7_stacks_and_queues/fish.cpp
// https://codility.com/programmers/task/fish/

int solution(vector<int> &A, vector<int> &B) {
    vector<int> downstream;
    int upsteamAlive = 0;
    for (size_t i = 0; i < A.size(); ++i)
    {
        if (B[i])
        {
            downstream.push_back(A[i]);
        }
        else
        {
            while (!downstream.empty() && A[i] > downstream.back())
            {
                downstream.pop_back();
            }
            if (downstream.empty())
                ++upsteamAlive;
        }
    }
    
    return upsteamAlive + downstream.size();
}

File: Codility/7_stacks_and_queues/fish.py
# // https://codility.com/programmers/task/fish/

def solution(A, B):
    downstream = []
    upstream_alive = 0
    
    for i in xrange(len(A)):
        if B[i]:
            downstream.append(A[i])
        else:
            while downstream and A[i] > downstream[len(downstream) - 1]:
                del downstream[len(downstream) - 1]
            
            if not downstream:
                upstream_alive += 1
                
    return upstream_alive + len(downstream)